expression/util.go

```diff
@@ -177,6 +177,31 @@ func ExtractEquivalenceColumns(result [][]Expression, exprs []Expression) [][]Ex
 	return result
 }
 
+// FindUpperBound looks for column < constant or column <= constant and returns both the column
+// and constant. It return nil, 0 if the expression is not of this form.
+// It is used by derived Top N pattern and it is put here since it looks like
+// a general purpose routine. Similar routines can be added to find lower bound as well.
+func FindUpperBound(expr Expression) (*Column, int64) {
+	scalarFunction, scalarFunctionOk := expr.(*ScalarFunction)
+	if scalarFunctionOk {
+		args := scalarFunction.GetArgs()
+		if len(args) == 2 {
+			col, colOk := args[0].(*Column)
+			constant, constantOk := args[1].(*Constant)
+			if colOk && constantOk && (scalarFunction.FuncName.L == ast.LT || scalarFunction.FuncName.L == ast.LE) {
+				value, valueOk := constant.Value.GetValue().(int64)
```

Comment:
Should we check the type of this constant here like `constant.GetType().EvalType() == types.ETInt && !mysql.HasUnsignedFlag(constant.GetType().GetFlag())`?

Reply:
The constant type is already checked by line 192 "value, valueOk := constant.Value.GetValue().(int64)". the unsigned check is done by checking if the value is >= 0 in the calling function. But, your suggestion for the unsigned looks cleaner.

Reply:
But, I would like to keep this routine generic to find upper bound. In this case, the consumer is checking if it is unsigned or not.

Reply:
Gotit